# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1495573933 25200
#      Tue May 23 14:12:13 2017 -0700
# Node ID 645d1f5f2cb2912322bf3e61de438b6a2809485e
# Parent  7c278572db66d16a6a2da04ef21073bd2dc73d55
pkg: warn on invalid or repeated lines in DESCRIPTION (bug #51090)

* get_description.m: Warn on lines with duplicate keywords and ignore the
contents instead of replacing.  Issue a warning for invalid lines.
* package.txi (The DESCRIPTION File): Fix doc of Depends keyword, multiple
lines are not accepted.

diff --git a/doc/interpreter/package.txi b/doc/interpreter/package.txi
--- a/doc/interpreter/package.txi
+++ b/doc/interpreter/package.txi
@@ -393,18 +393,17 @@ include dependencies on particular versi
 @example
 Depends: package (>= 1.0.0)
 @end example
 
 @noindent
 Possible operators are @code{<}, @code{<=}, @code{==}, @code{>=} or
 @code{>}.  If the part of the dependency in @code{()} is missing, any
 version of the package is acceptable.  Multiple dependencies can be
-defined either as a comma separated list or on separate @code{Depends}
-lines.
+defined as a comma separated list.
 
 @item License
 An optional short description of the used license (e.g., GPL version 3
 or newer).  This is optional since the file @file{COPYING} is mandatory.
 
 @item SystemRequirements
 These are the external install dependencies of the package and are not
 checked by the package manager.  This is here as a hint to the
diff --git a/scripts/pkg/private/get_description.m b/scripts/pkg/private/get_description.m
--- a/scripts/pkg/private/get_description.m
+++ b/scripts/pkg/private/get_description.m
@@ -40,27 +40,32 @@ function desc = get_description (filenam
       ## Continuation lines
       if (exist ("keyword", "var") && isfield (desc, keyword))
         desc.(keyword) = [desc.(keyword) " " deblank(line)];
       endif
     else
       ## Keyword/value pair
       colon = find (line == ":");
       if (length (colon) == 0)
-        disp ("skipping line");
+        warning ("pkg: skipping invalid line in DESCRIPTION file");
       else
         colon = colon(1);
         keyword = tolower (strtrim (line(1:colon-1)));
         value = strtrim (line (colon+1:end));
         if (length (value) == 0)
             fclose (fid);
             error ("The keyword '%s' of the package '%s' has an empty value",
                     keyword, desc.name);
         endif
-        desc.(keyword) = value;
+        if (isfield (desc, keyword))
+          warning ("pkg: duplicate keyword \"%s\" in DESCRIPTION, ignoring",
+                   keyword);
+        else
+          desc.(keyword) = value;
+        endif
       endif
     endif
     line = fgetl (fid);
   endwhile
   fclose (fid);
 
   ## Make sure all is okay.
   needed_fields = {"name", "version", "date", "title", ...
